File: Solutions/Subarray_Product_Less_Than_K/Subarray_Product_Less_Than_K.cpp
//Sliding window
class Solution
{

public:
    int numSubarrayProductLessThanK(vector<int> &nums, int k)
    {

        if (k == 0)
            return 0;

        int start = 0;
        int end = 0;
        int p = 1;
        int count = 0;

        while (end < nums.size())
        {
            p *= nums[end];

            while (start < nums.size() and p >= k)
            {
                p = p / nums[start];
                start++;
            }

            if (p < k)
                count += end - start + 1;

            end++;
        }

        return count;
    }
};

//Fastest
class Solution
{
public:
    int numSubarrayProductLessThanK(vector<int> &a, int k)
    {

        ios_base::sync_with_stdio(0);
        cin.tie(0);

        int p = 1;

        int st = 0, e = 0, ans = 0;

        while (e < a.size())
        {
            p = p * a[e];

            while (st <= e and p >= k)
            {
                p = p / a[st];
                st++;
            }

            ans += (e - st + 1);

            e++;
        }

        return ans;
    }
};

File: Solutions/Subarray_Product_Less_Than_K/Subarray_Product_Less_Than_K.cs
//Sliding window
public class Solution {
    public int NumSubarrayProductLessThanK(int[] nums, int k) {
    
    if(k==0) return 0;
    
    int start=0;
    int end=0;
    int p=1;
    int count=0;
    
    while(end<nums.Length)
    {
        p *= nums[end];
        
        while(start < nums.Length && p>=k)
        {
            p=p/nums[start];
            start++;
        }
        
        if(p<k)
        count+=end-start+1;
        
        end++;
        
    }
    
    return count;
}
}

//Most common
public class Solution {
    public int NumSubarrayProductLessThanK(int[] nums, int k) {
     
        if (nums == null || nums.Length == 0)
            return 0;
        
        int left = 0;
        var prod = 1;
        int upperBound;
        
        var count = 0;
        
        for(int right = 0; right < nums.Length; right++)
        {                    
            prod = prod * nums[right];
            while(prod >= k && left <= right)
            {                
                prod = prod/nums[left];    
                left++;
            }
            
            count += right - left + 1;
        }
        
        return count;
        
    }
}

//Fastest
public class Solution {
    public int NumSubarrayProductLessThanK(int[] nums, int k) {
     
        if (nums == null || nums.Length == 0)
            return 0;
        
        int left = 0;
        var prod = 1;
        int upperBound;
        
        var count = 0;
        
        for(int right = 0; right < nums.Length; right++)
        {                    
            prod *= nums[right];
            while(prod >= k && left <= right)
            {                
                prod = prod/nums[left];    
                left++;
            }
            
            count += right - left + 1;
        }
        
        return count;
        
    }
}